client-react/src/ApiHelpers/ArmHelper.ts

```diff
@@ -211,6 +216,187 @@ const MakeArmCall = async <T, U = T>(requestObject: ArmRequestObject<U>): Promis
+            headers: r.headers,
+            error: error,
+          },
+        });
+      } else {
+        retObj = pollAzureAsyncOperationForCompletion<T, U>(response, azureAsyncOperation, request);
+      }
+    }
+
+    return retObj;
+  });
+};
+
+const pollProvisioningStateForCompletion = <T, U = T>(request: ArmRequestObject<U>) => {
+  const retryAfter = 2000;
+  const setTelemetryHeader = request.commandName ? request.commandName + '-polling' : 'PollingAsyncResponse';
+
+  return delay(() => {
+    return MakeArmCall<T>({
+      resourceId: request.resourceId,
+      commandName: setTelemetryHeader,
+      method: 'GET',
+      apiVersion: request.apiVersion,
+    });
+  }, retryAfter).then(r => {
+    let retObj: Promise<HttpResponseObject<T>> | undefined = undefined;
+    if (r) {
+      const rSuccess = r.httpStatusCode < 300;
+
+      if (r.httpStatusCode === 200) {
+        retObj = Promise.resolve({
+          data: r.content,
+          metadata: {
+            success: rSuccess,
+            status: r.httpStatusCode,
+            headers: r.headers,
+            error: rSuccess ? null : r.content,
+          },
+        });
+      } else if (r.httpStatusCode === 201 || r.httpStatusCode === 202) {
+        const provisioningState = (<any>r.content)?.properties?.provisioningState;
+        if (provisioningState === ProvisioningState.Succeeded) {
+          retObj = Promise.resolve({
+            data: r.content,
+            metadata: {
+              success: rSuccess,
+              status: r.httpStatusCode,
+              headers: r.headers,
+            },
+          });
+        } else if (provisioningState === ProvisioningState.Failed) {
+          const error = r.content?.error || r.content;
+          retObj = Promise.resolve({
+            data: r.content,
+            metadata: {
+              success: rSuccess,
+              status: r.httpStatusCode,
+              headers: r.headers,
+              error: error,
+            },
+          });
+        } else {
+          retObj = pollProvisioningStateForCompletion(request);
+        }
+      } else {
+        const error = r.content || { code: r.httpStatusCode, message: null };
+        retObj = Promise.resolve({
+          data: r.content,
+          metadata: {
+            success: rSuccess,
+            status: r.httpStatusCode,
+            headers: r.headers,
+            error: error,
+          },
+        });
+      }
+    }
+
+    return retObj;
+  });
+};
+
+const getHeader = (headerToFind: string, headers: KeyValue<string>) => {
+  for (const key of Object.keys(headers)) {
+    if (Object.prototype.hasOwnProperty.call(headers, key) && key.toLowerCase() === headerToFind.toLowerCase()) {
+      return headers[key];
+    }
+  }
+};
+
+const delay = async (func: () => Promise<any>, ms: number = 3000) => {
+  const sleepPromise = new Promise(resolve => setTimeout(resolve, ms));
+  return await sleepPromise.then(func);
+};
```

Comment:
Avoid mixing `async`/`await` and `then`ables.
```suggestion
const delay = async (func: () => Promise<any>, ms = 3000) => {
  await new Promise(resolve => setTimeout(resolve, ms));
  return await func();
};
```